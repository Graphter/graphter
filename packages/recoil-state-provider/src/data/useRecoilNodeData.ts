import { useRecoilState } from "recoil";
import { propDataStore } from "../store/propDataStore";
import { NodeDataHook } from "@graphter/renderer-react";
import { nodeRendererStore } from "@graphter/renderer-react";
import { getValueByGlobalPath } from "@graphter/renderer-react";
import { useEffect } from "react";
import { PathMeta } from "@graphter/renderer-react";

/**
 * @param path
 */
export const useRecoilNodeData: NodeDataHook = (
  path,
  config,
  originalTreeData
  ) => {

  let parentMeta: PathMeta | null = null
  let parentSetMeta: ((meta: PathMeta) => void) | null = null

  if(path.length > 2){
    // We can do this conditional because path is expected to be stable for this instance of the hook
    const parentState = propDataStore.get(path.slice(0, -1))
    if(!parentState) throw new Error('Parent state is expected to always have been stored before a child is processed')
    const parent = useRecoilState(parentState)
    parentMeta = parent[0]
    parentSetMeta = parent[1]
  }

  if(!propDataStore.has(path)){
    // Store meta for current path

    propDataStore.set(generatePathMeta())

    if(path.length > 2 && parentMeta && parentSetMeta){
      // Add path to parent
      parentMeta = {
        ...parentMeta,
        childPaths: [
          ...parentMeta.childPaths,
          path
        ]
      }
      parentSetMeta(parentMeta)
    }
  }

  const propDataState = propDataStore.get(path)
  if(!propDataState) throw new Error('Should have state by now')

  const [ pathMeta, setPathMeta ] = useRecoilState(propDataState)

  useEffect(() => {
    if(!pathMeta.nodes.length){
      // Looks like the path meta hasn't been initialised yet - could happen if another node has requested it before it's nodes have rendered
      setPathMeta(generatePathMeta())
    }
  })

  return [ pathMeta.nodes.find(node => node.config.id === config.id)?.internalData, (newData: any) => {
    setPathMeta({
      ...pathMeta,
      nodes: pathMeta.nodes.map(node => {
        return node.config.id === config.id ?
          {
            ...node,
            internalData: newData
          } :
          node
      })
    })
  } ]

  function generatePathMeta(): PathMeta {
    const rendererRegistration = nodeRendererStore.get(config.type)
    const fallbackValue = rendererRegistration.createFallbackDefaultValue ?
      rendererRegistration.createFallbackDefaultValue(config, path, (path) => getValueByGlobalPath(originalTreeData, path)) :
      null
    return {
      path,
      nodes: [
        {
          config,
          internalData: rendererRegistration.initialiser ?
            rendererRegistration.initialiser(originalTreeData, config, path) :
            getValueByGlobalPath(originalTreeData, path, fallbackValue),
          rendererRegistration
        }
      ],
      childPaths: []
    }
  }
}
