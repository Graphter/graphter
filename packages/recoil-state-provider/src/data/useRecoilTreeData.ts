import { useRecoilValue } from "recoil";
import { TreeDataHook } from "@graphter/renderer-react";
import { PathSegment } from "@graphter/core";
import branchDataStore from "../store/branchDataStore";

export const useRecoilTreeData: TreeDataHook = <T>(path: Array<PathSegment>, depth?: number) => {
  return useRecoilValue<T>(branchDataStore.get<T>(path, depth))
}