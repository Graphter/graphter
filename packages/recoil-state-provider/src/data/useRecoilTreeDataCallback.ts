import { useRecoilCallback } from "recoil";
import treeDataStore from "../store/treeDataStore";
import { TreeDataCallbackHook } from "@graphter/renderer-react";

export const useRecoilTreeDataCallback: TreeDataCallbackHook = (fn, config, path) => {
  return useRecoilCallback(({snapshot}) => async () => {
    fn(await snapshot.getPromise(treeDataStore.getDescendentData(config, path)))
  })
}