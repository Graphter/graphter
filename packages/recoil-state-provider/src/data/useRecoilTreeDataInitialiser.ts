import { propDataStore } from "../store/propDataStore";
import { TreeDataInitialiserHook } from "@graphter/renderer-react";
import { getTreeMeta } from "@graphter/renderer-react";
import { PathSegment } from "@graphter/core";
import { getValue } from "@graphter/renderer-react"
import { createDefault, nodeRendererStore } from "@graphter/renderer-react";

const NO_MATCH = 'no-match-ac437190-4001-4fe3-bf1f-8ad03366d1dd'

export const useRecoilTreeDataInitialiser: TreeDataInitialiserHook = () => {
  return (config, path, committed = true, originalTreeData) => {

    const getNodeValue = (path: Array<PathSegment>) => {
      const result = getValue(originalTreeData, path, NO_MATCH)
      if (result === NO_MATCH) return
      return result
    }

    const localPath = path.slice(2)
    const treeMeta = getTreeMeta(config, localPath, getNodeValue)

    for (const nodeMeta of treeMeta) {
      if (!nodeMeta.path.length) continue
      const globalPath = [ ...path.slice(0, 2), ...nodeMeta.path ]
      if (propDataStore.has(globalPath)) continue
      const rendererReg = nodeRendererStore.get(nodeMeta.config.type)
      if(rendererReg.initialiser){
        const initialData = rendererReg.initialiser(originalTreeData, nodeMeta.config, nodeMeta.path)
        propDataStore.set(globalPath, initialData)
      } else {
        let initialData = getValue(originalTreeData, nodeMeta.path, NO_MATCH)
        if (initialData === NO_MATCH) {
          initialData = createDefault(
            nodeMeta.config,
            rendererReg.createFallbackDefaultValue ?
              rendererReg.createFallbackDefaultValue(nodeMeta.config, globalPath, getNodeValue) :
              null
          )
        }
        propDataStore.set(globalPath, initialData)
      }



    }

  }

}