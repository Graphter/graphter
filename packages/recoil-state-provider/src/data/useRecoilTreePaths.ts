import { useRecoilValue, useRecoilValueLoadable } from 'recoil';
import { PathSegment } from '@graphter/core';
import treeDataStore from '../store/treeDataStore';
import { TreePathsHook } from '@graphter/renderer-react';

export const useRecoilTreePaths:TreePathsHook = (config, path) => {
  return useRecoilValue<Array<Array<PathSegment>>>(treeDataStore.getBranchPaths(config, path))
}