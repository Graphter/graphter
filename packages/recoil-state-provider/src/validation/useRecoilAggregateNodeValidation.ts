import { useRecoilValue } from 'recoil';
import {
  PathSegment,
} from '@graphter/core';
import validationDataStore from '../store/validationDataStore';
import { AggregateNodeValidationHook } from "@graphter/renderer-react";

export const useRecoilAggregateNodeValidation: AggregateNodeValidationHook = (
  paths: Array<Array<PathSegment>>
) => {
  const aggregateValidationState = validationDataStore.getAll(paths)
  return useRecoilValue(aggregateValidationState)
}