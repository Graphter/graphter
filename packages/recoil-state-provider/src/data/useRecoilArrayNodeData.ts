import { nanoid } from "nanoid";
import { useRecoilNodeData } from "./useRecoilNodeData";
import { propDataStore } from "../store/propDataStore";
import { ArrayNodeDataHook } from "@graphter/renderer-react";
import { useState } from "react";

/**
 * @param path
 */
export const useRecoilArrayNodeData: ArrayNodeDataHook = (
  path,
) => {

  const [ arrayItems ] = useRecoilNodeData(path)

  const [ childIds, setChildIds ] = useState(arrayItems.map(() => nanoid()))

  return {
    childIds,
    removeItem: (index: number) => {
      const clone = [...childIds]
      clone.splice(index, 1)
      const childPath = [ ...path, index ]
      propDataStore.remove(childPath)
      setChildIds(clone)
    },
    commitItem: (index: number) => {
      const clone = [...childIds]
      clone.splice(index, 0, nanoid())
      const childPath = [ ...path, index ]
      propDataStore.commitItem(childPath)
      setChildIds(clone)
    }
  }
}