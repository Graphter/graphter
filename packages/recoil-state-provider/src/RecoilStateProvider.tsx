import { RecoilRoot } from "recoil";
import React, { ReactNode } from "react";
import { NodeValidatorRegistration } from "@graphter/core";
import { useRecoilNodeValidation } from "./validation/useRecoilNodeValidation";
import { useRecoilAggregateNodeValidation } from "./validation/useRecoilAggregateNodeValidation";
import { useRecoilNodeData } from "./data/useRecoilNodeData";
import { useRecoilTreeDataCallback } from "./data/useRecoilTreeDataCallback";
import { useRecoilTreePaths } from "./data/useRecoilTreePaths";
import { useRecoilArrayNodeData } from "./data/useRecoilArrayNodeData";
import { NodeDataProvider, NodeValidationProvider } from "@graphter/renderer-react";
import { useRecoilTreeDataInitialiser } from "./data/useRecoilTreeDataInitialiser";
import { useRecoilTreeData } from "./data/useRecoilTreeData";

export interface RecoilStateProvider {
  instanceId: string | number
  validatorRegistry: Array<NodeValidatorRegistration>
  children?: ReactNode
}

export function RecoilStateProvider(
  {
    instanceId,
    validatorRegistry,
    children
  }: RecoilStateProvider
) {
  return (
    <RecoilRoot>
      <NodeValidationProvider
        instanceId={instanceId}
        nodeValidationHook={useRecoilNodeValidation}
        aggregateNodeValidationHook={useRecoilAggregateNodeValidation}
        validatorRegistry={validatorRegistry}
      >
        <NodeDataProvider
          instanceId={instanceId}
          treeDataHook={useRecoilTreeData}
          treeDataInitialiserHook={useRecoilTreeDataInitialiser}
          nodeDataHook={useRecoilNodeData}
          treeDataCallbackHook={useRecoilTreeDataCallback}
          treePathsHook={useRecoilTreePaths}
          arrayNodeDataHook={useRecoilArrayNodeData}
        >
          {children}
        </NodeDataProvider>
      </NodeValidationProvider>
    </RecoilRoot>
  )
}

export default RecoilStateProvider