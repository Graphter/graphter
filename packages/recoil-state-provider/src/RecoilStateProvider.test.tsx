import { render } from "@testing-library/react";
import RecoilStateProvider from "./RecoilStateProvider";
import React from "react";
import { NodeDataProvider, NodeValidationProvider } from "@graphter/renderer-react";
import { useRecoilAggregateNodeValidation } from "./validation/useRecoilAggregateNodeValidation";
import { useRecoilNodeValidation } from "./validation/useRecoilNodeValidation";
import { NodeValidatorRegistration } from "@graphter/core";
import { useRecoilTreeDataInitialiser } from "./data/useRecoilTreeDataInitialiser";
import { useRecoilNodeData } from "./data/useRecoilNodeData";
import { useRecoilTreeData } from "./data/useRecoilTreeData";
import { useRecoilTreeDataCallback } from "./data/useRecoilTreeDataCallback";
import { useRecoilTreePaths } from "./data/useRecoilTreePaths";
import { useRecoilArrayNodeData } from "./data/useRecoilArrayNodeData";

const NodeValidationProviderMock = NodeValidationProvider as jest.Mock<any>
const NodeDataProviderMock = NodeDataProvider as jest.Mock<any>

describe('<RecoilStateProvider />', () => {
  it('should correctly set up Recoil state management', () => {
    const validatorRegistry: Array<NodeValidatorRegistration> = []
    const { container } = render(
      <RecoilStateProvider instanceId={'some-id'} validatorRegistry={validatorRegistry}>
        <div>Some child</div>
      </RecoilStateProvider>
    )
    expect(container).toMatchSnapshot()
    expect(NodeValidationProviderMock).toHaveBeenCalled()
    const nodeValidationProviderProps = NodeValidationProviderMock.mock.calls[0][0]
    expect(nodeValidationProviderProps.instanceId).toBe('some-id')
    expect(nodeValidationProviderProps.nodeValidationHook).toBe(useRecoilNodeValidation)
    expect(nodeValidationProviderProps.aggregateNodeValidationHook).toBe(useRecoilAggregateNodeValidation)
    expect(nodeValidationProviderProps.validatorRegistry).toBe(validatorRegistry)
    expect(Object.keys(nodeValidationProviderProps)).toMatchSnapshot()
    expect(NodeDataProviderMock).toHaveBeenCalled()
    const nodeDataProviderProps = NodeDataProviderMock.mock.calls[0][0]
    expect(nodeDataProviderProps.instanceId).toBe('some-id')
    expect(nodeDataProviderProps.treeDataInitialiserHook).toBe(useRecoilTreeDataInitialiser)
    expect(nodeDataProviderProps.nodeDataHook).toBe(useRecoilNodeData)
    expect(nodeDataProviderProps.treeDataHook).toBe(useRecoilTreeData)
    expect(nodeDataProviderProps.treeDataCallbackHook).toBe(useRecoilTreeDataCallback)
    expect(nodeDataProviderProps.treePathsHook).toBe(useRecoilTreePaths)
    expect(nodeDataProviderProps.arrayNodeDataHook).toBe(useRecoilArrayNodeData)
    expect(Object.keys(nodeDataProviderProps)).toMatchSnapshot()

  })
})
export {}