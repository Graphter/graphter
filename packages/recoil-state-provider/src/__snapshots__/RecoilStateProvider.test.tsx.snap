// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RecoilStateProvider /> should correctly set up Recoil state management 1`] = `
<div>
  <div>
    <div>
      <div>
        Some child
      </div>
    </div>
  </div>
</div>
`;

exports[`<RecoilStateProvider /> should correctly set up Recoil state management 2`] = `
Array [
  "instanceId",
  "nodeValidationHook",
  "aggregateNodeValidationHook",
  "validatorRegistry",
  "children",
]
`;

exports[`<RecoilStateProvider /> should correctly set up Recoil state management 3`] = `
Array [
  "instanceId",
  "treeDataHook",
  "treeDataInitialiserHook",
  "nodeDataHook",
  "treeDataCallbackHook",
  "treePathsHook",
  "arrayNodeDataHook",
  "children",
]
`;
